vispy/visuals/pseudo_spheres.py

```diff
@@ -0,0 +1,204 @@
+# Copyright (c) Vispy Development Team. All Rights Reserved.
+# Distributed under the (new) BSD License. See LICENSE.txt for more info.
+# -----------------------------------------------------------------------------
+"""Pseudo spheres (shaded discs) visual"""
+
+import numpy as np
+
+from ..color import ColorArray
+from ..gloo import VertexBuffer
+from .shaders import Function, Variable
+from .visual import Visual
+
+
+vert = """
+uniform float u_px_scale;
+uniform float u_scale;
+uniform vec3 u_light_position;
+
+attribute vec3 a_position;
+attribute vec4 a_color;
+attribute float a_radius;
+
+varying vec4 v_color;
+varying float v_radius;
+varying vec3 v_light_direction;
+varying float v_depth;
+varying float v_depth_middle;
+
+void main (void) {
+    v_color = a_color;
+    v_radius = a_radius * u_scale * u_px_scale;
+    v_light_direction = normalize(u_light_position);
+
+    vec4 pos = vec4(a_position, 1.0);
+    vec4 fb_pos = $visual_to_framebuffer(pos);
+    gl_Position = $framebuffer_to_render(fb_pos);
+
+    // Get the framebuffer z direction relative to this sphere in visual coords
+    vec4 z = $framebuffer_to_visual(fb_pos + vec4(0, 0, -100, 0));
+    z = (z/z.w - pos) / 100;
+
+    // Get the depth of the sphere in its middle (+ radius)
+    vec4 depth_z = $framebuffer_to_render($visual_to_framebuffer(pos + normalize(z) * a_radius));
+    v_depth_middle = gl_Position.z / gl_Position.w - depth_z.z / depth_z.w;
+
+    gl_PointSize = v_radius;
+}
+"""
+
+
+frag = """
+uniform vec3 u_light_color;
+uniform float u_light_ambient;
+
+varying vec4 v_color;
+varying float v_radius;
+varying vec3 v_light_direction;
+varying float v_depth;
+varying float v_depth_middle;
+
+void main()
+{
+    // discard radius 0 spheres
+    if (v_radius <= 0.)
+        discard;
+
+    // discard fragments outside of disc
+    vec2 texcoord = gl_PointCoord * 2.0 - vec2(1.0);
+    float x = texcoord.x;
+    float y = texcoord.y;
+    float d = 1.0 - x*x - y*y;
+    if (d <= 0)
+        discard;
+
+    float z = sqrt(d);
+    vec3 normal = vec3(x,y,z);
+
+    // Diffuse color
+    float diffuse = dot(v_light_direction, normal);
+    // clamp, because 0 < theta < pi/2
+    diffuse = clamp(diffuse, 0.0, 1.0);
+    vec3 diffuse_color = u_light_ambient + u_light_color * diffuse;
+
+    // Specular color
+    //   reflect light wrt normal for the reflected ray, then
+    //   find the angle made with the eye
+    vec3 eye = vec3(0, 0, -1);
+    float specular = dot(reflect(v_light_direction, normal), eye);
+    specular = clamp(specular, 0.0, 1.0);
+    // raise to the material's shininess, multiply with a
+    // small factor for spread
+    specular = pow(specular, 80);
+    vec3 specular_color = u_light_color * specular;
+
+    gl_FragColor = vec4(v_color.rgb * diffuse_color + specular_color, v_color.a);
+
+    // TODO: why was it 0.5? And why is 0.36 better?
+    gl_FragDepth = gl_FragCoord.z - 0.36 * z * v_depth_middle;
```

Comment:
I believe it is 0.5 here because `v_depth_middle` is set by `a_radius`. Since `a_radius` is being used to set `gl_PointSize`, `a_radius` actually ends up being the diameter (see ref below). Thus, the 0.5 was to convert to radius.

https://www.khronos.org/registry/OpenGL-Refpages/gl2.1/xhtml/glPointSize.xml

Reply:
Thanks, solved together with the change in coordinate system!